### PR TITLE
[Profile] `az login`: Fall back to device code flow in GitHub Codespaces

### PR DESCRIPTION
**Pull request by sinedied**
**Related command**

`az login`

**Description**

This PR fixes the long standing issue of `az login` not working in GitHub Codespaces (https://github.com/Azure/azure-cli/issues/20315).
For that it uses a simple workaround, the same used in the gh CLI: if the `CODESPACES=true` environnement variable is set, it falls back to using a device code login.

**Testing Guide**

1. In a bash terminal: run `az login` -> by default, it should run the web browser login
2. Run `CODESPACES=true az login` -> it should automatically fallback to device code login. You can also simply run `az login` in GitHub Codespaces :)

**History Notes**

[Core] `az login`: Add fallback to device code in GitHub Codespaces (#20315)

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

**Comment by sinedied** [comment]
Tests are finally green 🙂 
Sorry for that, I did not succeed in the running the tests locally, `azdev test` command wasn't working for me

**Comment by sinedied** [comment]
@jiasli The PR is ready for review, could you have a look?

This will also solve the same issue in [AZD CLI](https://github.com/Azure/azure-dev) as it uses AZ CLI under the hood

**Comment by jiasli** [comment]
@sinedied, thanks for your contribution. Device code has some known limitations with Conditional Access policies, so we prefer users to use auth code flow.

Also, as users can use `--use-device-code` to force `az login` to use device code, this PR may not be urgent at the moment. We will evaluate this PR with our PMs when time permits. Thanks for your understanding.

File: src/azure-cli-core/azure/cli/core/util.py [hunk 1]
```diff
@@ -760,6 +760,10 @@ def is_windows():
     return platform_name == 'windows'
 
 
+def is_github_codespaces():
+    return os.environ.get('CODESPACES', None) == 'true'
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/util.py:764]: It's better to put [the document of `CODESPACES`](https://docs.github.com/en/codespaces/developing-in-a-codespace/default-environment-variables-for-your-codespace) here. 

`dict.get` already uses `None` as the `default`: https://docs.python.org/3/library/stdtypes.html#dict.get.

```suggestion
    # https://docs.github.com/en/codespaces/developing-in-a-codespace/default-environment-variables-for-your-codespace
    return os.environ.get('CODESPACES') == 'true'
```

File: src/azure-cli-core/azure/cli/core/tests/test_profile.py [hunk 2]
```diff
@@ -418,6 +420,29 @@ def test_login_in_cloud_shell(self, msi_auth_mock, create_subscription_client_mo
...
+        mock_subscription_client = mock.MagicMock()
+        mock_subscription_client.tenants.list.return_value = [TenantStub(self.tenant_id)]
+        mock_subscription_client.subscriptions.list.return_value = [deepcopy(self.subscription1_raw)]
+        create_subscription_client_mock.return_value = mock_subscription_client
+
+        storage_mock = {'subscriptions': None}
+        profile = Profile(cli_ctx=cli, storage=storage_mock)
+        subs = profile.login(True, None, None, False, None, use_device_code=False, allow_no_subscriptions=False)
+
+        # assert
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/tests/test_profile.py:443]: We may check if `azure.cli.core.auth.identity.Identity.login_with_device_code` is really called:

```suggestion
        # assert
        login_with_device_code_mock.assert_called_once()
```

File: src/azure-cli-core/azure/cli/core/tests/test_profile.py [hunk 3]
```diff
@@ -418,6 +420,29 @@ def test_login_in_cloud_shell(self, msi_auth_mock, create_subscription_client_mo
...
         self.assertEqual(s['id'], self.id1.split('/')[-1])
 
+    @mock.patch('azure.cli.core._profile.SubscriptionFinder._create_subscription_client', autospec=True)
+    @mock.patch('azure.cli.core.auth.identity.Identity.get_user_credential', autospec=True)
+    @mock.patch('azure.cli.core._profile.can_launch_browser', autospec=True, return_value=True)
+    @mock.patch('azure.cli.core.auth.identity.Identity.login_with_device_code', autospec=True)
+    @mock.patch('azure.cli.core._profile.is_github_codespaces', autospec=True, return_value=True)
+    def test_login_in_codespaces(self, is_github_codespaces_mock, 
+                                 login_with_device_code_mock, can_launch_browser_mock,
+                                 get_user_credential_mock, create_subscription_client_mock):
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/tests/test_profile.py:430]: It's better to align the patch order with other test methods:

```suggestion
    @mock.patch('azure.cli.core._profile.SubscriptionFinder._create_subscription_client', autospec=True)
    @mock.patch('azure.cli.core.auth.identity.Identity.get_user_credential', autospec=True)
    @mock.patch('azure.cli.core.auth.identity.Identity.login_with_device_code', autospec=True)
    @mock.patch('azure.cli.core._profile.can_launch_browser', autospec=True, return_value=True)
    @mock.patch('azure.cli.core._profile.is_github_codespaces', autospec=True, return_value=True)
    def test_login_in_codespaces(self, is_github_codespaces_mock, can_launch_browser_mock,
                                 login_with_device_code_mock, get_user_credential_mock,
                                 create_subscription_client_mock):
```

File: src/azure-cli-core/azure/cli/core/tests/test_profile.py [hunk 4]
```diff
@@ -597,7 +622,8 @@ def test_find_subscriptions_in_vm_with_msi_user_assigned_with_res_id(self, creat
     @mock.patch('azure.cli.core.auth.identity.Identity.get_user_credential', autospec=True)
     @mock.patch('azure.cli.core.auth.identity.Identity.login_with_auth_code', autospec=True)
     @mock.patch('azure.cli.core._profile.can_launch_browser', autospec=True, return_value=True)
-    def test_login_no_subscription(self, can_launch_browser_mock,
+    @mock.patch('azure.cli.core._profile.is_github_codespaces', autospec=True, return_value=False)
+    def test_login_no_subscription(self, is_github_codespaces_mock, can_launch_browser_mock,
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/tests/test_profile.py:626]: We don't support running these tests in GitHub Codespaces, so I don't think it is necessary to patch `is_github_codespaces` in these test methods.

**Comment by jiasli** [comment]
@sinedied, I understand you might be busy working on other tasks and don't have the bandwidth on this PR, so I will close this PR and continue the work in https://github.com/Azure/azure-cli/pull/29826. Thanks for your understanding.

**Comment by sinedied** [comment]
@jiasli sorry for the delay, I missed the notification last week. My last update I received by mail was that you weren't going to merge it, but if this has changed please reopen this PR and I'll get it updated today

**Comment by jiasli** [comment]
I have verified the PR works as expected in GitHub Codespaces:

image

**Review by jiasli — DISMISSED** [review]

**Review by bebound — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Review by jiasli — APPROVED** [review]

**Comment by sneakyflint** [comment]
Hello - wondering if there would be any consideration given to either reverting this change or adding a CLI flag to bypass this check? `az login` has worked well in GitHub Codespaces for quite some time in our enterprise - the automatic port forwarding service picks up the port used by the callback server and forwards with private scope:
image

The exception is when using Codespaces on Web which does not support binding ports to 127.0.0.1 - though this could be worked around by allowing the end user to pass a replacement value for `localhost` when calling az login.

We have worked around with `CODESPACES=false az login` for now but wanted to ensure this was flagged.

**Comment by jiasli** [comment]
> `az login` has worked well in GitHub Codespaces for quite some time in our enterprise - the automatic port forwarding service picks up the port used by the callback server and forwards with private scope:

Are you [using a local VS Code](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-github-codespaces-in-visual-studio-code) to connect to Codespaces?

I can indeed get it working as `localhost:45439` is mapped to `127.0.0.1:45439`.

image

> The exception is when using Codespaces on Web which does not support binding ports to 127.0.0.1 - though this could be worked around by allowing the end user to pass a replacement value for `localhost` when calling az login.

No, this can't be done. This explained in https://github.com/Azure/azure-cli/issues/20315#issuecomment-1898372497.

**Comment by jiasli** [comment]
BTW, I encountered a weird issue that the local VS Code somtimes cannot launch a web browser (Edge, Chrome). Instead, it launches a text-based browser in the terminal.

image

This can be triggered by directly running:

```
$ python3 -c "import webbrowser; webbrowser.open('https://login.microsoft.com/')"
Unable to connect to VS Code server: Error in request.
Error: connect ENOENT /tmp/vscode-ipc-51da01ec-3078-41e1-944c-868de003c9c7.sock
    at PipeConnectWrap.afterConnect [as oncomplete] (node:net:1607:16) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'connect',
  address: '/tmp/vscode-ipc-51da01ec-3078-41e1-944c-868de003c9c7.sock'
}
```

**Comment by sneakyflint** [comment]
> > `az login` has worked well in GitHub Codespaces for quite some time in our enterprise - the automatic port forwarding service picks up the port used by the callback server and forwards with private scope:
> 
> Are you [using a local VS Code](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-github-codespaces-in-visual-studio-code) to connect to Codespaces?
> 
> I can indeed get it working as VS Code automatically maps **`localhost` on the remote container** as **`127.0.0.1` on the local machine**.
> 
> image
> 
> This is also documented at https://docs.github.com/en/codespaces/developing-in-a-codespace/forwarding-ports-in-your-codespace#using-command-line-tools-and-rest-clients-to-access-ports
> 
> > If you forward a private port from the VS Code desktop application, your application will also be available at a localhost port such as `127.0.0.1:4000`.
> 
> > The exception is when using Codespaces on Web which does not support binding ports to 127.0.0.1 - though this could be worked around by allowing the end user to pass a replacement value for `localhost` when calling az login.
> 
> No, this can't be done. This explained in [#20315 (comment)](https://github.com/Azure/azure-cli/issues/20315#issuecomment-1898372497).

Yes, we are using local VSCode. Makes sense re: the redirect URL issue and we have been able to get by w/ using VSCode desktop or reverting to the device code method manually. 



> BTW, I encountered a weird issue that the local VS Code somtimes cannot launch a web browser (Edge, Chrome). Instead, it launches a text-based browser in the terminal.
> 
> image
> 
> This can be triggered by directly running:
> 
> ```
> $ python3 -c "import webbrowser; webbrowser.open('https://login.microsoft.com/')"
> Unable to connect to VS Code server: Error in request.
> Error: connect ENOENT /tmp/vscode-ipc-51da01ec-3078-41e1-944c-868de003c9c7.sock
>     at PipeConnectWrap.afterConnect [as oncomplete] (node:net:1607:16) {
>   errno: -2,
>   code: 'ENOENT',
>   syscall: 'connect',
>   address: '/tmp/vscode-ipc-51da01ec-3078-41e1-944c-868de003c9c7.sock'
> }
> ```

Just tried this on my end and it immediately opened a browser window, when running on `mcr.microsoft.com/devcontainers/base:ubuntu-22.04`. Perhaps an upstream issue with VSCode server/client?

**Comment by tallaxes** [comment]
FWIW, we are using Codespaces with VSCode extensively, and the old workflow (automatic port forwarding + web browser) was working fine for us. With this fix, we are now discovering that login via device code flow gets somewhat different permissions (in our case breaking `az role assignment create`, possibly due to lack of permissions for ARG from an "unmanaged" VM?). So for now will be using `CODESPACES=false az login` as a workaround🙂

**Pull request merged**